##  Analyzing Used Car Listings on eBay Kleinanzeigen

In this project, we will be working on a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally [scraped](https://en.wikipedia.org/wiki/Classified_advertising) and uploaded to Kaggle by user orgesleka.

The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

There have been a few modifications from the original dataset:

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- When the ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing.
- price - The listed selling price of the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The type of vehicle.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The type of transmission.
- powerPS - The power of the car in PS.
- model - The car model name.
- odometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

Let's start by importing the required libraries, then load the dataset using pandas.

In [1]:
import pandas as pd
import numpy as np
np.set_printoptions(suppress=True)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

In [2]:
autos = pd.read_csv('../Datasets/autos.csv', encoding='latin1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Cleaning Column Names

> observations:

- The dataset contains 20 columns, most of which are stored as strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and change some of the column names based on the data dictionary in order to be more descriptive.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename({'yearOfRegistration': 'registration_year',
             'monthOfRegistration': 'registration_month',
             'notRepairedDamage': 'unrepaired_damage',
             'dateCreated': 'ad_created'},
            axis=1, inplace=True)

In [7]:
autos.columns

mylist = []

for col in autos.columns:
    new = ''
    for char in col:
        if char.isupper():
            new += '_' + char.lower()
        else:
            new += char
    mylist.append(new)

In [8]:
autos.columns = mylist

In [9]:
mylist

['date_crawled',
 'name',
 'seller',
 'offer_type',
 'price',
 'abtest',
 'vehicle_type',
 'registration_year',
 'gearbox',
 'power_p_s',
 'model',
 'odometer',
 'registration_month',
 'fuel_type',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'nr_of_pictures',
 'postal_code',
 'last_seen']

In [10]:
autos.rename({'power_p_s': 'power_ps', 'abtest':'ab_test',
            'nr_of_pictures': 'num_photos'},
           axis=1, inplace=True)

In [11]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [12]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

##  Initial Exploration and Cleaning

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [13]:
# Check descriptive statistics
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox    power_ps  model   odometer  \
count         44905        50000.00000    47320 50000.00000  47242      50000   
unique            8                NaN        2         NaN    245         13   
top       limousine                NaN  manuell         NaN   golf  150,000km   
freq          12859                NaN    36993         NaN   4024      32424   
mean            NaN         2005.07328      NaN   116.35592    NaN        NaN   
std             NaN          105.71281      NaN   209.21663    NaN        NaN   
min             NaN         1000.00000      NaN     0.00000    NaN        NaN   
25%             NaN         1999.00000      NaN    70.00000    NaN        NaN   
50%             NaN         2003.00000      NaN   105.00000    NaN        NaN   
75%             NaN         2008.00000      NaN   150.00000    NaN        NaN   
max             NaN         9999.00000      NaN 17700.00000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count          50000.00000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean               5.72336       NaN         NaN               NaN   
std                3.71198       NaN         NaN               NaN   
min                0.00000       NaN         NaN               NaN   
25%                3.00000       NaN         NaN               NaN   
50%                6.00000       NaN         NaN               NaN   
75%                9.00000       NaN         NaN               NaN   
max               12.00000       NaN         NaN               NaN   

                 ad_created  num_photos  postal_code            last_seen  
count                 50000 50000.00000  50000.00000                50000  
unique                   76         NaN          NaN                39481  
top     2016-04-03 00:00:00         NaN          NaN  2016-04-07 06:17:27  
freq                   1946         NaN          NaN                    8  
mean                    NaN     0.00000  50813.62730                  NaN  
std                     NaN     0.00000  25779.74796                  NaN  
min                     NaN     0.00000   1067.00000                  NaN  
25%                     NaN     0.00000  30451.00000                  NaN  
50%                     NaN     0.00000  49577.00000                  NaN  
75%                     NaN     0.00000  71540.00000                  NaN  
max                     NaN     0.00000  99998.00000                  NaN

Our initial observations:

- There are a number of text columns where all (or nearly all) of the values are the same:
    - `seller`
    - `offer_type`
- The `num_photos` column looks odd, we'll need to investigate this further.

In [14]:
autos['num_photos'].value_counts()

num_photos
0    50000
Name: count, dtype: int64

It looks like the `num_photos` column has `0` for every column.  We'll drop this column, plus the other two we noted as mostly one value.

In [15]:
autos = autos.drop(['seller', 'offer_type', 'num_photos'], axis=1)

There are two columns, `price` and `odometer`, which are numeric values with extra characters being stored as text.  We'll clean and convert these.

In [16]:
autos['price'] = (autos['price']
                  .str.replace('$', '')
                  .str.replace(',', '')
                  .astype(int)
                 )

autos['odometer'] = (autos['odometer']
                     .str.replace('km', '')
                     .str.replace(',', '')
                     .astype(int)
                    )

In [17]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [18]:
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [19]:
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

## Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [20]:
autos[['price', 'odometer_km']].describe()

price  odometer_km
count    50000.00000  50000.00000
mean      9840.04376 125732.70000
std     481104.38050  40042.21171
min          0.00000   5000.00000
25%       1100.00000 125000.00000
50%       2950.00000 150000.00000
75%       7200.00000 150000.00000
max   99999999.00000 150000.00000

In [21]:
autos['odometer_km'].value_counts()

odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64

In [22]:
autos['price'].value_counts().head()

price
0       1421
500      781
1500     734
2500     643
1000     639
Name: count, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field.  Additionally, there are more high mileage than low mileage vehicles.

In [23]:
autos['price'].unique().shape

(2357,)

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.


There are 1,421 cars listed with $0 price - given that this is only 2% of the cars, we might consider removing these rows.  The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [24]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64

In [25]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64

There are a number of listings with prices below $30, including about 1,500 at $0.  There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1.  We will keep the \$1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [29]:
autos = autos[autos['price'].between(1, 351000)]

In [31]:
autos['price'].max()

350000

In [32]:
autos['price'].min()

1

In [34]:
autos['price'].describe()

count    48565.00000
mean      5888.93559
std       9059.85475
min          1.00000
25%       1200.00000
50%       3000.00000
75%       7490.00000
max     350000.00000
Name: price, dtype: float64

## Exploring the Date Columns

Let's now move on to the date columns and understand the date range the data covers.

There are 5 columns that represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [35]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

You'll notice that the first 10 characters represent the day (e.g. 2016-03-12). To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

In [36]:
autos['date_crawled'].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object

In [50]:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index())

date_crawled
2016-03-05   0.02533
2016-03-06   0.01404
2016-03-07   0.03601
2016-03-08   0.03330
2016-03-09   0.03309
2016-03-10   0.03218
2016-03-11   0.03257
2016-03-12   0.03692
2016-03-13   0.01567
2016-03-14   0.03655
2016-03-15   0.03428
2016-03-16   0.02961
2016-03-17   0.03163
2016-03-18   0.01291
2016-03-19   0.03478
2016-03-20   0.03789
2016-03-21   0.03737
2016-03-22   0.03299
2016-03-23   0.03222
2016-03-24   0.02934
2016-03-25   0.03161
2016-03-26   0.03220
2016-03-27   0.03109
2016-03-28   0.03486
2016-03-29   0.03410
2016-03-30   0.03369
2016-03-31   0.03183
2016-04-01   0.03369
2016-04-02   0.03548
2016-04-03   0.03861
2016-04-04   0.03649
2016-04-05   0.01310
2016-04-06   0.00317
2016-04-07   0.00140
Name: proportion, dtype: float64

In [60]:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values()
)

date_crawled
2016-04-07   0.00140
2016-04-06   0.00317
2016-03-18   0.01291
2016-04-05   0.01310
2016-03-06   0.01404
2016-03-13   0.01567
2016-03-05   0.02533
2016-03-24   0.02934
2016-03-16   0.02961
2016-03-27   0.03109
2016-03-25   0.03161
2016-03-17   0.03163
2016-03-31   0.03183
2016-03-10   0.03218
2016-03-26   0.03220
2016-03-23   0.03222
2016-03-11   0.03257
2016-03-22   0.03299
2016-03-09   0.03309
2016-03-08   0.03330
2016-04-01   0.03369
2016-03-30   0.03369
2016-03-29   0.03410
2016-03-15   0.03428
2016-03-19   0.03478
2016-03-28   0.03486
2016-04-02   0.03548
2016-03-07   0.03601
2016-04-04   0.03649
2016-03-14   0.03655
2016-03-12   0.03692
2016-03-21   0.03737
2016-03-20   0.03789
2016-04-03   0.03861
Name: proportion, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [61]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

last_seen
2016-03-05   0.00107
2016-03-06   0.00432
2016-03-07   0.00539
2016-03-08   0.00741
2016-03-09   0.00960
2016-03-10   0.01067
2016-03-11   0.01238
2016-03-12   0.02378
2016-03-13   0.00890
2016-03-14   0.01260
2016-03-15   0.01588
2016-03-16   0.01645
2016-03-17   0.02809
2016-03-18   0.00735
2016-03-19   0.01583
2016-03-20   0.02065
2016-03-21   0.02063
2016-03-22   0.02137
2016-03-23   0.01853
2016-03-24   0.01977
2016-03-25   0.01921
2016-03-26   0.01680
2016-03-27   0.01565
2016-03-28   0.02086
2016-03-29   0.02234
2016-03-30   0.02477
2016-03-31   0.02378
2016-04-01   0.02279
2016-04-02   0.02492
2016-04-03   0.02520
2016-04-04   0.02448
2016-04-05   0.12476
2016-04-06   0.22181
2016-04-07   0.13195
Name: proportion, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [52]:
(autos['ad_created']
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())

ad_created
2015-06-11   0.00002
2015-08-10   0.00002
2015-09-09   0.00002
2015-11-10   0.00002
2015-12-05   0.00002
               ...  
2016-04-03   0.03886
2016-04-04   0.03686
2016-04-05   0.01182
2016-04-06   0.00325
2016-04-07   0.00126
Name: proportion, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [49]:
autos['registration_year'].describe()

count   48565.00000
mean     2004.75542
std        88.64389
min      1000.00000
25%      1999.00000
50%      2004.00000
75%      2008.00000
max      9999.00000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car.  Looking at this column, we note some odd values.  The minimum value is `1000`, long before cars were invented and the maximum is `9999`, many years into the future.

## Dealing with Incorrect Registration Year Data

One thing that stands out from the exploration we did in the above is that the registration_year column contains some odd values:

- The minimum value is 1000, before cars were invented
- The maximum value is 9999, many years into the future

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.  Determining the earliest valid year is more difficult.  Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values.  Let's determine what percentage of our data has invalid values in this column:

In [65]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.

In [84]:
# Many ways to select rows in a dataframe that fall within a value range for a column.
# Using `Series.between()` is one way.
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000   0.06761
2005   0.06289
1999   0.06206
2004   0.05790
2003   0.05782
2006   0.05720
2001   0.05647
2002   0.05326
1998   0.05062
2007   0.04878
Name: proportion, dtype: float64

## Exploring Price by Brand

In [72]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [85]:
autos["brand"].value_counts(normalize=True)

brand
volkswagen       0.21126
bmw              0.11004
opel             0.10758
mercedes_benz    0.09646
audi             0.08657
ford             0.06990
renault          0.04715
peugeot          0.02984
fiat             0.02564
seat             0.01827
skoda            0.01641
nissan           0.01527
mazda            0.01519
smart            0.01416
citroen          0.01401
toyota           0.01270
hyundai          0.01003
sonstige_autos   0.00981
volvo            0.00915
mini             0.00876
mitsubishi       0.00823
honda            0.00784
kia              0.00707
alfa_romeo       0.00664
porsche          0.00613
suzuki           0.00593
chevrolet        0.00570
chrysler         0.00351
dacia            0.00263
daihatsu         0.00251
jeep             0.00227
subaru           0.00214
land_rover       0.00210
saab             0.00165
jaguar           0.00156
daewoo           0.00150
trabant          0.00139
rover            0.00133
lancia           0.00107
lada             0.

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [86]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In [87]:
mean_price_by_brand = {}

for brand in common_brands:
    
    # get the mean price per brand
    mean_price = autos[autos['brand'] == brand]['price'].mean()
    
    mean_price_by_brand[brand] = mean_price

In [88]:
mean_price_by_brand

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

Of the top 5 brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

## Storing Aggregate Data in a DataFrame

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

In [93]:
top_6_brands = autos['brand'].value_counts(normalize=True).head(6).index

In [96]:
autos.head(2)

date_crawled                                        name  price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   

   ab_test vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   

   odometer_km  registration_month fuel_type    brand unrepaired_damage  \
0       150000                   3       lpg  peugeot              nein   
1       150000                   6    benzin      bmw              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08

In [98]:
mean_price_by_brand = {}
mean_mileage_by_brand = {}

for brand in top_6_brands:
    mean_price = (
        autos[autos['brand'] == brand]['price'].mean()
    )
    mean_mileage = (
        autos[autos['brand'] == brand]['odometer_km'].mean()
    )
    mean_price_by_brand[brand] = mean_price
    mean_mileage_by_brand[brand] = mean_mileage
    

In [99]:
mean_price_by_brand

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

In [100]:
mean_mileage_by_brand

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056}

In [105]:
series = pd.Series(mean_price_by_brand)

In [111]:
df = pd.DataFrame(series, columns=['mean_price'])
df['mean_mileage'] = mean_mileage_by_brand

In [112]:
df

mean_price  mean_mileage
volkswagen     5402.41026  128707.15879
bmw            8332.82052  132572.51314
opel           2975.24194  129310.03584
mercedes_benz  8628.45037  130788.36331
audi           9336.68745  129157.38679
ford           3749.46951  124266.01287

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

In this project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps for you to consider:

Data cleaning next steps:
Identify categorical data that uses german words, translate them and map the values to their english counterparts
Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
See if there are particular keywords in the name column that you can extract as new columns